preview/services/store.py

```diff
@@ -27,4 +27,356 @@
 
 
 """
-import boto3
\ No newline at end of file
+import io
+import binascii
+from datetime import datetime
+from hashlib import md5
+from base64 import b64encode
+from typing import IO, Tuple, Optional, Dict, Any
+
+from typing_extensions import TypedDict, Literal
+from pytz import UTC
+from flask import Flask
+import boto3
+import botocore
+from botocore.config import Config
+from botocore.exceptions import ClientError
+
+from arxiv.base import logging
+from arxiv.base.globals import get_application_global, get_application_config
+
+from ..domain import Content, Preview, Metadata
+
+logger = logging.getLogger(__name__)
+
+
+class HeadResponse(TypedDict):
+    """Expected response to HEAD object requests."""
+
+    ETag: str
+    LastModified: datetime
+    ContentLength: int
+
+
+class GetResponse(TypedDict):
+    """Expected response to GET object requests."""
+
+    ETag: str
+    LastModified: datetime
+    ContentLength: int
+    Body: IO[bytes]
+
+
+class StreamMonitor(io.BytesIO):
+    """Wraps a stream to calculate checksum and size as the stream is read."""
+
+    def __init__(self, stream: IO[bytes]) -> None:
+        """Initialize md5 hash and size."""
+        self._stream = stream
+        self._md5 = md5()
+        self.size_bytes = 0
+        super().__init__()
+
+    def seekable(self) -> Literal[False]:
+        """Indicate that this is a non-seekable stream."""
+        return False
+
+    def readable(self) -> Literal[True]:
+        """Indicate that it *is* a readable stream."""
+        return True
+
+    def read(self, size: Optional[int] = -1) -> bytes:
+        """Update the hash and size as the stream is read."""
+        if size is None:
+            size = -1
+        chunk = self._stream.read(size)
+        self._md5.update(chunk)
+        self.size_bytes += len(chunk)
+        return chunk
+
+    @property
+    def checksum(self) -> str:
+        """Get the base64-encoded MD5 hash of the stream content."""
+        return self._md5.hexdigest()
```

Comment:
Nah, S3 calculates its checksums differently than what we are doing for the source checksums. But this makes me think that I should review docstrings to make sure that I'm not telling a different tale somewhere...

Reply:
In that case, the docstring here just needs to be revised? (since hashlib's `md5.hexdigest()` doesn't appear to do b64-encoding behind the scenes)

Reply:
I was overthinking this. In 31c6fb3 using the b64 of the digest so that we're consistent with everything else. When using the ETag from S3 I just convert that from hex to the b64+digest form.